api/nodes/permissions.py

```diff
@@ -63,6 +65,36 @@ def has_object_permission(self, request, view, obj):
             has_auth = parent_node.can_edit(auth)
             return has_auth
 
+
+class ContributorOrPublicForComments(permissions.BasePermission):
+
+    def has_object_permission(self, request, view, obj):
+        assert isinstance(obj, (Node, Comment, CommentReport)), 'obj must be a Node, Comment or CommentReport, got {}'.format(obj)
+        auth = get_user_auth(request)
+        if isinstance(obj, CommentReport):
+            comment_id = request.parser_context['kwargs']['comment_id']
+            comment = Comment.load(comment_id)
+            if not comment:
+                raise NotFound
+            node = comment.node
+        elif isinstance(obj, Comment):
+            comment = obj
+            node = obj.node
+        elif isinstance(obj, Node):
+            node = obj
+
+        view_name = view.get_view_name()
+        is_report_view = (view_name == 'Comment Reports List' or view_name == 'Comment Report Detail')
```

Comment:
This gets around the circular import. Not sure on the policy of scoped imports. I've only ever used them in circumstances where they solved a circular import. @abought, Thoughts?

```python
class ContributorOrPublicForComments(permissions.BasePermission):

    def has_object_permission(self, request, view, obj):
        from api.comments import views as comment_views

        assert isinstance(obj, (Node, Comment, CommentReport)), 'obj must be a Node, Comment or CommentReport, got {}'.format(obj)
        auth = get_user_auth(request)
        if isinstance(obj, CommentReport):
            comment_id = request.parser_context['kwargs']['comment_id']
            comment = Comment.load(comment_id)
            if not comment:
                raise NotFound
            node = comment.node
        elif isinstance(obj, Comment):
            comment = obj
            node = obj.node
        elif isinstance(obj, Node):
            node = obj

        # view_name = view.get_view_name()
        # is_report_view = (view_name == 'Comment Reports List' or view_name == 'Comment Report Detail')
        is_report_view = isinstance(view, comment_views.CommentReportDetail) or isinstance(view, comment_views.CommentReportsList)

        if not is_report_view and request.method in permissions.SAFE_METHODS:
            return node.is_public or node.can_view(auth)
        else:
            if isinstance(view, comment_views.CommentDetail):
                return comment.user._id == auth.user._id
            else:
                return node.can_comment(auth)
```